[上一页：1.2.1终止与恢复](/ThinkingInJava原版/异常捕获/1.2.1终止与恢复.md)
# 创建自定义异常
- [自定义有参构造器](#自定义无参构造器的异常)
- [自定义有参构造器](#自定义有参构造器)

java自身提供的那点异常类说实话根本不够使的。所以我们可以自定义异常。但是写的时候
你也不能瞎写，必须继承自已有的异常类。

##### 自定义无参构造器的异常
```
/**
 * 自定义的exception！
 * 对于异常来说，最最最重要的是，类名！
 */
class SimpleException extends Exception { }

public class InheritingExceptions {
    public void f() throws SimpleException {
        System.out.println("Throw SimpleException from f()");
        throw new SimpleException();
    }

    public static void main(String[] args) {
        InheritingExceptions inheritingExceptions = new InheritingExceptions();
        try {
            inheritingExceptions.f();
        } catch (SimpleException si) {
            System.out.println("Caught it!");
        }
    }
}
```
对于异常而言，最为重要的是起一个能够看得懂的名字！

##### 自定义有参构造器
接下来将会介绍一下，exception的一些比较有用的方法：

1 Exception自带含参构造器

2 e.printStackTrace(); // 这个方法会把堆栈打印出来！
```
/**
 * 纸质书 252页
 * 异常，有参构造器
 */
class MyException extends Exception {
    public MyException() {
    }

    public MyException(String msg) {
        super(msg);
    }
}

public class FullConstructors {
    public static void f() throws MyException {
        System.out.println("Throwing MyException from f()");
        throw new MyException();
    }

    public static void g() throws MyException {
        System.out.println("Throwing MyException from g()");
        throw new MyException("Originated in g()");
    }

    public static void main(String[] args) {
        try {
            f();
        } catch (MyException e) {
            e.printStackTrace(); // 这个方法会把堆栈打印出来！
        }

        try {
            g();
        } catch (MyException e) {
            e.printStackTrace();
        }
    }
}
```
[下一页：1.3.1异常与记录日志](/ThinkingInJava原版/异常捕获/1.3.1异常与记录日志.md)